x/streamer/keeper/distribute.go

```diff
@@ -0,0 +1,174 @@
+func (k Keeper) getToDistributeCoinsFromStreams(streams []types.Stream) sdk.Coins {
+	coins := sdk.Coins{}
+	distributed := sdk.Coins{}
+
+	for _, stream := range streams {
+		coins = coins.Add(stream.Coins.Sort()...)
+		distributed = distributed.Add(stream.DistributedCoins...)
+	}
+	return coins.Sub(distributed...)
+}
+
+// getToDistributeCoinsFromIterator utilizes iterator to return a list of streams.
+// From these streams, coins that have not yet been distributed are returned
+func (k Keeper) getToDistributeCoinsFromIterator(ctx sdk.Context, iterator db.Iterator) sdk.Coins {
+	return k.getToDistributeCoinsFromStreams(k.getStreamsFromIterator(ctx, iterator))
+}
+
+// getDistributedCoinsFromIterator utilizes iterator to return a list of streams.
+// From these streams, coins that have already been distributed are returned
+func (k Keeper) getDistributedCoinsFromIterator(ctx sdk.Context, iterator db.Iterator) sdk.Coins {
+	return k.getDistributedCoinsFromStreams(k.getStreamsFromIterator(ctx, iterator))
+}
+
+// moveUpcomingStreamToActiveStream moves a stream that has reached it's start time from an upcoming to an active status.
+func (k Keeper) moveUpcomingStreamToActiveStream(ctx sdk.Context, stream types.Stream) error {
+	// validation for current time and distribution start time
+	if ctx.BlockTime().Before(stream.StartTime) {
+		return fmt.Errorf("stream is not able to start distribution yet: %s >= %s", ctx.BlockTime().String(), stream.StartTime.String())
+	}
+
+	timeKey := getTimeKey(stream.StartTime)
+	if err := k.deleteStreamRefByKey(ctx, combineKeys(types.KeyPrefixUpcomingStreams, timeKey), stream.Id); err != nil {
+		return err
+	}
+	if err := k.addStreamRefByKey(ctx, combineKeys(types.KeyPrefixActiveStreams, timeKey), stream.Id); err != nil {
+		return err
+	}
+	return nil
+}
+
+// moveActiveStreamToFinishedStream moves a stream that has completed its distribution from an active to a finished status.
+func (k Keeper) moveActiveStreamToFinishedStream(ctx sdk.Context, stream types.Stream) error {
+	timeKey := getTimeKey(stream.StartTime)
+	if err := k.deleteStreamRefByKey(ctx, combineKeys(types.KeyPrefixActiveStreams, timeKey), stream.Id); err != nil {
+		return err
+	}
+	if err := k.addStreamRefByKey(ctx, combineKeys(types.KeyPrefixFinishedStreams, timeKey), stream.Id); err != nil {
+		return err
+	}
+	return nil
+}
+
+// distributeInternal runs the distribution logic for a stream, and adds the sends to
+// the distrInfo struct. It also updates the stream for the distribution.
+func (k Keeper) distributeInternal(
+	ctx sdk.Context, stream types.Stream,
+) (sdk.Coins, error) {
+	totalDistrCoins := sdk.NewCoins()
+	remainCoins := stream.Coins.Sub(stream.DistributedCoins...)
+	remainEpochs := uint64(stream.NumEpochsPaidOver - stream.FilledEpochs)
+
+	for _, coin := range remainCoins {
+		epochAmt := coin.Amount.Quo(sdk.NewInt(int64(remainEpochs)))
```

Comment:
If the amount of tokens is lower than epochs  (i.e 29/30) we won't divide the remaining balance. What is the desired logic here?  seems like it's just gonna stay and "die" with the stream?

Reply:
the stream gonna get to finished state, so those token will be calculated in the "free" token of the streamer... they won't be stuck on the finished stream